Op overload parser

ajreynol opened the pull request:
This commit adds support for operator overloading in the smt2 parser. It is the remaining portion of #1117 (with improvements), whose first phase was merged in #1154.

The changes required:

- The Smt2 grammar for term had to be split into term and termNonVariable to accommodate reading variable expressions with type casts via AS_TOK.

- The commit required some refactoring of the Parser class so that some query functions take expressions instead of names as arguments.

- We have to distinguish when we allow for symbols to be overloaded (see argument "doOverload" throughout). Currently, top-level declarations (including datatype declarations) and definitions with name clashes are treated as overloaded symbols in the smt2 parser only. Other things like local variable declarations are not treated as overloaded.

barrettcw commented on src/parser/parser.h:
```diff
@@ -322,6 +324,7 @@ class CVC4_PUBLIC Parser {
    * @return the variable expression
+   * Only returns a variable if its name is not overloaded.
    */
   Expr getVariable(const std::string& name);
```
Also, this needs a better name.  We have getVariableExpressoin and getVariable - the method names should distinguish how these are different - maybe this one should be lookupVariable?

barrettcw commented on src/parser/parser.cpp:
```diff
@@ -92,11 +93,59 @@ Expr Parser::getFunction(const std::string& name) {
+    // the variable is overloaded, try with type if the type exists
+    if(!t.isNull()) {
+      // if we support annotations for function types, this will update to 
+      // separate t into ( argument types, return type )
```
You're saying that if we later decide to support annotations, then...
right?
say "if we decide later..." to make it clear

barrettcw commented on src/parser/parser.cpp:
```diff
@@ -92,11 +93,59 @@ Expr Parser::getFunction(const std::string& name) {
+  // now, post-process the expression
+  assert( !expr.isNull() );
+  if(isDefinedFunction(expr)) {
+    expr = d_exprManager->mkExpr(CVC4::kind::APPLY, expr);
```
I don't understand what's happening here.  expr is a function definition?  Why are we wrapping it in an APPLY?

barrettcw commented on src/parser/parser.cpp:
```diff
@@ -92,11 +93,59 @@ Expr Parser::getFunction(const std::string& name) {
+  }else{
+    Type te = expr.getType();
+    if(te.isConstructor() && ConstructorType(te).getArity() == 0) {
+      // immediately turn it into an apply constructor
```
So I'm guessing this is just our internal convention?  Nullary constructors are represented by applying the constructor to nothing?  Maybe remind the reader of this convention here.

barrettcw commented on src/parser/parser.cpp:
```diff
@@ -121,16 +170,17 @@ size_t Parser::getArity(const std::string& sort_name) {
 }
 
 /* Returns true if name is bound to a function-like thing (function,
  * constructor, tester, or selector). */
```
This comment should be in the .h file

barrettcw commented on src/parser/parser.cpp:
```diff
@@ -214,15 +265,19 @@ std::vector<Expr> Parser::mkBoundVars(const std::vector<std::string> names,
-                            bool levelZero) {
-  d_symtab->bindDefinedFunction(name, val, levelZero);
+                            bool levelZero, bool doOverload) {
+  d_symtab->bindDefinedFunction(name, val, levelZero, doOverload);
```
What if this returns false?

ajreynol commented on src/parser/parser.h:
```diff
@@ -330,10 +333,37 @@ class CVC4_PUBLIC Parser {
+   * Returns a expression, given a name and a type. 
+   *
+   * This is the same as above but where the name has been type cast to t. 
+   * A null type for t indicates that there was no type cast.
```
One is when we parse [name] and the other is called when we parser (as [name] [t]).  I'll remove the comment (but internally the first function calls the second).

barrettcw replied:
My suggestion would be to have getExpressionForNameAndType have all of the details (including that a null type indicates no type cast) and then indicate that getExpressionForName is a shortcut for calling getExpressionForNameAndType with a null type.

ajreynol commented:
I've addressed the points in the review.
I changed the name getVariableExpression to getExpressionForName, and kept the name of other functions like getVariable the same.

barrettcw commented on src/parser/smt2/Smt2.g:
```diff
@@ -1788,6 +1796,20 @@ symbolicExpr[CVC4::SExpr& sexpr]
+: termNonVariable[expr, expr2]
+    /* a variable */
+  | symbol[name,CHECK_DECLARED,SYM_VARIABLE]
+    { expr = PARSER_STATE->getExpressionForName(name); }
```
What happen if expr is NULL here?

ajreynol replied:
Expression can't be null here.  I noticed that assert(...)'s are not too common in this file I suppose since CVC4 assertions aren't available, that's why I typically haven't been adding more assertions to ANTLR files.  Regardless, I'd prefer them so I'll add assert( !expr.isNull() );

barrettcw commented on src/parser/smt2/Smt2.g:
```diff
@@ -1852,17 +1876,28 @@ term[CVC4::Expr& expr, CVC4::Expr& expr2]
+      if(readVariable) {
+        Trace("parser-overloading") << "Getting variable expression of type " << name << " with type " << type << std::endl;
+        // get the variable expression for the type
+        f = PARSER_STATE->getExpressionForNameAndType(name, type); 
```
What happens if f is Null?

ajreynol replied:
Same as above
